In [1]:
import numpy as np
import scipy as sp
import matplotlib.pyplot as plt

AU_TO_EV = 27.2113961317875

# SoftCoulomb 1D

In [2]:
# сглаженный кулон
def potential(x, k, d):
    return -k/np.sqrt(x**2+d**2)

In [3]:
# затравочная волновая функция основного состояния
def psi0(x, b):
    return (b/np.pi)**(1/4) * np.exp(-b*x**2 / 2)

In [4]:
# энергия основного состояния
def K0(x):
    return sp.special.kn(0, x)
    
def K1(x):
    return sp.special.kn(1, x)
    
def E0(b, k, d):
    return b/4 - k*np.sqrt(b/np.pi)*np.exp(d**2*b/2) * K0(d**2*b/2)

In [5]:
# уравнение, минимизирующее энергию (далее не используется)
def dE0_db(b, k, d):
    t = d**2 * b
    a = 1/k
    return np.exp(t/2)/np.sqrt(t) * (
        (1+t)*K0(t/2) - t*K1(t/2)
    ) - np.sqrt(np.pi)/2 * a/d

In [6]:
# минимизация энергии
def energy_minimization(k, d):
    minimization_results = sp.optimize.minimize(lambda b: E0(b, k, d), x0=1)
    E_min = minimization_results["fun"]
    b_min = minimization_results["x"][0]
    return b_min, E_min

In [7]:
# подбор параметра d для заданной энергии
# dmin, dmax -- диапазон d
def fit_d(E0, k, dmin, dmax):
    delta_energy = lambda d: energy_minimization(k, d)[1] - E0
    return sp.optimize.bisect(delta_energy, dmin, dmax)

## He1 основное состояние

In [10]:
# рассмотрим He1
z = 2
Ip = 54.41776 / AU_TO_EV
d_fitted = fit_d(-Ip, z, 0.01, 1.5)
b_min, E_min = energy_minimization(z, d_fitted)
print("d_fitted = ", d_fitted)
print("b_min    = ", b_min)
print("E_min    = ", E_min, "au = ", E_min*AU_TO_EV, "eV")

d_fitted =  0.7030159450553491
b_min    =  1.2156924048233868
E_min    =  -1.9998150677904902 au =  -54.4177599999645 eV


## H1 основное состояние

In [24]:
# рассмотрим H1
z = 1
Ip = 0.7
d_fitted = fit_d(-Ip, z, 0.000001, 1)
b_min, E_min = energy_minimization(z, d_fitted)
print("d_fitted = ", d_fitted)
print("b_min    = ", b_min)
print("E_min    = ", E_min, "au = ", E_min*AU_TO_EV, "eV")

d_fitted =  0.9400473308748792
b_min    =  0.49416871449028177
E_min    =  -0.7000000000001223 au =  -19.047977292254576 eV


## Ar18 основное состояние

In [8]:
# рассмотрим Ar18
k = 18
d_fitted = fit_d(-162.6608564501, k, 0.001, 0.1)
b_min, E_min = energy_minimization(k, d_fitted)
print("d_fitted = ", d_fitted)
print("b_min    = ", b_min)
print("E_min    = ", E_min, "au = ", E_min*AU_TO_EV, "eV")

d_fitted =  0.07773024924787751
b_min    =  99.06619189087772
E_min    =  -162.66085645276254 au =  -4426.229000071944 eV


Получаем, что волновая функция основного состояния Ar18:
$$
\psi_0(x) = \left(\dfrac{b}{\pi}\right)^{1/4}\exp\left(-\dfrac{b x^2}{2}\right)
$$
где $b=99.06619189087772$

In [10]:
# x = np.load("/home/denis/RustSSFM/RSSFM2D/src/out/dim1/x0.npy")
# psi_initial = psi0(x, b_min)
# np.save("/home/denis/RustSSFM/RSSFM2D/src/out/dim1/psi_initial.npy", psi_initial.astype(np.complex64))

In [60]:
# протестируем мои параметры из SSFM расчета
k = 18
d = 0.07818

b_min, E_min = energy_minimization(k, d)
print("d     = ", d)
print("b_min = ", b_min)
print("E_min = ", E_min)

d     =  0.07818
b_min =  98.36704765237765
E_min =  -161.86708013130024


Хорошо сходятся!